File: integrations/paperclip-bridge/src/schemas.ts
import { z } from "zod";

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export const ServiceType = z.enum([
  "website_landing",
  "website_ecommerce",
  "website_portfolio",
  "webapp_saas",
  "webapp_internal_tool",
  "webapp_dashboard",
  "automation_email",
  "automation_crm",
  "automation_data_pipeline",
  "automation_scraper",
  "automation_reporting",
]);
export type ServiceType = z.infer<typeof ServiceType>;

export const Priority = z.enum(["low", "normal", "high", "urgent"]);
export type Priority = z.infer<typeof Priority>;

export const ProjectStatus = z.enum([
  "received",
  "scoping",
  "in_progress",
  "review",
  "delivered",
  "revision_requested",
  "completed",
  "cancelled",
]);
export type ProjectStatus = z.infer<typeof ProjectStatus>;

// ---------------------------------------------------------------------------
// Project Request (received FROM Aura via TPCP)
// ---------------------------------------------------------------------------

export const ProjectRequest = z.object({
  request_id: z.string().uuid(),
  lead_id: z.string(),
  company_name: z.string().max(255),
  contact_email: z.string().email(),
  contact_name: z.string().optional(),
  contact_phone: z.string().optional(),
  service_type: ServiceType,
  requirements: z.string(),
  budget_usd: z.number().min(0),
  priority: Priority.default("normal"),
  deadline_iso: z.string().datetime().optional(),
  research_notes: z.string().optional(),
  existing_website: z.string().url().optional(),
  industry: z.string().optional(),
  company_size: z
    .enum(["solo", "2-10", "11-50", "51-200", "200+"])
    .optional(),
  metadata: z.record(z.unknown()).default({}),
});
export type ProjectRequest = z.infer<typeof ProjectRequest>;

// ---------------------------------------------------------------------------
// Project Status Update (sent TO Aura via TPCP)
// ---------------------------------------------------------------------------

export const ProjectStatusUpdate = z.object({
  request_id: z.string().uuid(),
  lead_id: z.string(),
  paperclip_ticket_id: z.string().optional(),
  status: ProjectStatus,
  message: z.string().optional(),
  progress_pct: z.number().int().min(0).max(100).default(0),
  deliverable_url: z.string().url().optional(),
  deliverable_type: z
    .enum(["github_repo", "deployed_url", "zip_download", "notion_doc", "other"])
    .optional(),
  agent_notes: z.string().optional(),
  cost_usd: z.number().optional(),
  updated_at: z.string().datetime().default(() => new Date().toISOString()),
});
export type ProjectStatusUpdate = z.infer<typeof ProjectStatusUpdate>;

// ---------------------------------------------------------------------------
// Paperclip REST API types
// ---------------------------------------------------------------------------

export interface PaperclipTicket {
  id: string;
  title: string;
  description: string;
  status: string;
  assigneeId?: string;
  companyId: string;
  goalId?: string;
  metadata?: Record<string, unknown>;
}

export interface PaperclipCreateTicketRequest {
  title: string;
  description: string;
  companyId: string;
  goalId?: string;
  assigneeId?: string;
  priority?: "low" | "medium" | "high" | "urgent";
  metadata?: Record<string, unknown>;
}

File: integrations/aura-bridge/schemas.py
"""
Pydantic models for Aura-App <-> Paperclip TPCP communication.
Matches the JSON schema in ../shared/schemas.json
"""
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any
from uuid import UUID, uuid4

from pydantic import BaseModel, EmailStr, Field, HttpUrl


class ServiceType(str, Enum):
    WEBSITE_LANDING = "website_landing"
    WEBSITE_ECOMMERCE = "website_ecommerce"
    WEBSITE_PORTFOLIO = "website_portfolio"
    WEBAPP_SAAS = "webapp_saas"
    WEBAPP_INTERNAL_TOOL = "webapp_internal_tool"
    WEBAPP_DASHBOARD = "webapp_dashboard"
    AUTOMATION_EMAIL = "automation_email"
    AUTOMATION_CRM = "automation_crm"
    AUTOMATION_DATA_PIPELINE = "automation_data_pipeline"
    AUTOMATION_SCRAPER = "automation_scraper"
    AUTOMATION_REPORTING = "automation_reporting"


class Priority(str, Enum):
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"
    URGENT = "urgent"


class ProjectStatus(str, Enum):
    RECEIVED = "received"
    SCOPING = "scoping"
    IN_PROGRESS = "in_progress"
    REVIEW = "review"
    DELIVERED = "delivered"
    REVISION_REQUESTED = "revision_requested"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class CompanySize(str, Enum):
    SOLO = "solo"
    SMALL = "2-10"
    MEDIUM = "11-50"
    LARGE = "51-200"
    ENTERPRISE = "200+"


class ProjectRequest(BaseModel):
    """Sent by Aura-App to Paperclip when a lead qualifies for service delivery."""

    request_id: UUID = Field(default_factory=uuid4)
    lead_id: str = Field(..., description="Aura internal lead ID")
    company_name: str = Field(..., max_length=255)
    contact_email: str = Field(..., description="Primary contact email")
    contact_name: str | None = None
    contact_phone: str | None = None
    service_type: ServiceType
    requirements: str = Field(..., description="What the client needs")
    budget_usd: float = Field(..., ge=0, description="Client budget in USD")
    priority: Priority = Priority.NORMAL
    deadline_iso: datetime | None = None
    research_notes: str | None = Field(
        None, description="Aura's research: pain points, tech stack, opportunities"
    )
    existing_website: str | None = None
    industry: str | None = None
    company_size: CompanySize | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class ProjectStatusUpdate(BaseModel):
    """Received from Paperclip as the project progresses."""

    request_id: UUID
    lead_id: str
    paperclip_ticket_id: str | None = None
    status: ProjectStatus
    message: str | None = None
    progress_pct: int = Field(0, ge=0, le=100)
    deliverable_url: str | None = None
    deliverable_type: str | None = None
    agent_notes: str | None = None
    cost_usd: float | None = None
    updated_at: datetime = Field(default_factory=datetime.utcnow)
